core/zoe/zoe/zoe.js

```diff
@@ -0,0 +1,207 @@
+          adminState.setQuantitiesFor(offerIds, reallocation);
+        },
+
+        /**
+         * The governing contract can "complete" an offer to remove it
+         * from the ongoing governing contract and resolve the
+         * `result` promise with the player's payouts (either winnings or
+         * refunds). Because Zoe only allows for reallocations that
+         * conserve rights and are 'offer safe', we don't need to do
+         * those checks at this step and can assume that the
+         * invariants hold.
+         * @param  {array} offerIds - an array of offerIds
+         */
+        complete: offerIds => {
+          const quantities = readOnlyState.getQuantitiesFor(offerIds);
+          const amounts = toAmountMatrix(readOnlyState.getAssays(), quantities);
+          const payments = makePayments(adminState.getPurses(), amounts);
+          const results = adminState.getResultsFor(offerIds);
+          results.map((result, i) => result.res(payments[i]));
+          adminState.removeOffers(offerIds);
+        },
+
+        /**
+         *  The governing contract can create an empty offer and get
+         *  the associated offerId. This allows the governing contract
+         *  to use this offer slot for recordkeeping. For instance, to
+         *  represent a pool, the governing contract can create an
+         *  empty offer and then reallocate other quantities to this offer.
+         */
+        escrowEmptyOffer: async () => {
+          // attenuate the authority by not passing along the result
+          // promise object and only passing the offerId
+          const { offerId } = await escrowEmptyOffer(
+            adminState,
+            readOnlyState.getAssays(),
+            readOnlyState.getStrategies(),
+          );
+          return offerId;
+        },
+        /**
+         *  The governing contract can also create a real offer and
+         *  get the associated offerId, bypassing the seat and receipt
+         *  creation. This allows the governing contract to make
+         *  offers on the users' behalf, as happens in the
+         *  `addLiquidity` step of the `autoswap` contract.
+         */
+        escrowOffer: async (offerDesc, offerPayments) => {
+          // attenuate the authority by not passing along the result
+          // promise object and only passing the offerId
+          const { offerId } = await escrowOffer(
+            adminState,
+            readOnlyState.getStrategies(),
+            offerDesc,
+            offerPayments,
+          );
+          return offerId;
+        },
+
+        burnEscrowReceipt: async escrowReceipt => {
+          const amount = await escrowReceiptIssuer.burnAll(escrowReceipt);
+          return amount.quantity;
+        },
+
+        // read-only, side-effect-free access below this line:
+        makeEmptyQuantities: () =>
+          makeEmptyQuantities(readOnlyState.getStrategies()),
+        getStrategies: readOnlyState.getStrategies,
```

Comment:
Possibilities:

...readOnlyState?

Treat readOnlyState as a trait and reuse the trait composition pattern?

provide a getReadOnlyState() method whose result is the readOnlyState object itself.

I think I prefer the last.

Reply:
What kind of pros and cons are you thinking of compared to how it is now?